src/main/java/org/opentripplanner/middleware/triptracker/ManageLegTraversal.java

```diff
@@ -58,28 +58,46 @@ private static boolean canUseLeg(Leg leg) {
     }
 
     /**
-     * Get the expected leg by comparing the current time against the start and end time of each leg.
+     * Get the expected leg by first checking to see if two points on a leg contain the current position. If there is a
+     * match, return this leg, if not simply return the leg that is nearest to the current position.
      */
     @Nullable
-    public static Leg getExpectedLeg(Instant timeNow, Itinerary itinerary) {
+    public static Leg getExpectedLeg(Coordinates position, Itinerary itinerary) {
         if (canUseTripLegs(itinerary)) {
-            for (int i = 0; i < itinerary.legs.size(); i++) {
-                Leg leg = itinerary.legs.get(i);
-                if (i == 0 &&
-                    leg.mode.equalsIgnoreCase("walk") &&
-                    leg.startTime != null &&
-                    timeNow.isBefore(leg.startTime.toInstant())) {
-                    // If the first leg is a walk leg and the traveler has decided to start the trip early.
-                    return leg;
+            Leg leg = getContainingLeg(position, itinerary);
+            return (leg != null) ? leg : getNearestLeg(position, itinerary);
+        }
+        return null;
+    }
+
+    /**
+     * Get the leg that is nearest to the current position.
+     */
+    private static Leg getNearestLeg(Coordinates position, Itinerary itinerary) {
+        double shortestDistance = Double.MAX_VALUE;
+        Leg nearestLeg = null;
+        for (int i = 0; i < itinerary.legs.size(); i++) {
+            Leg leg = itinerary.legs.get(i);
+            for (Coordinates positionOnLeg : getAllLegPositions(leg)) {
+                double distance = getDistance(positionOnLeg, position);
+                if (distance < shortestDistance) {
+                    nearestLeg = leg;
+                    shortestDistance = distance;
```

Comment:
I forsee problems doing this at transit centers. If a bus goes down a ~200m path to get to the transit center and the next leg goes back the other direction how will we know which to match. Not sure I know how to fix this, but should we maybe add a comment that this might be an issue?

Reply:
This PR focuses on walking, but you are raising a good point for when we work on transit legs.

Reply:
There is an attempt to get the leg that contains the current position first (`getContainingLeg`) and then to fallback on this method (`getNearestLeg`) if unsuccessful. But, (having written all that) you are right that the wrong leg could be selected. I'll add a comment for future PRs to consider and potentially address.